mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala

```diff
@@ -378,6 +378,22 @@ object PowerIterationClustering extends Logging {
       logInfo(s"$msgPrefix: delta = $delta.")
       diffDelta = math.abs(delta - prevDelta)
       logInfo(s"$msgPrefix: diff(delta) = $diffDelta.")
+
+      if (math.abs(diffDelta) < tol) {
+        /**
+         * Power Iteration fails to converge if absolute value of top 2 eigen values are equal,
+         * but with opposite sign. The resultant vector flip-flops between two vectors.
+         * We should give an exception, if we detect the failure of the convergence of the
+         * power iteration
+         */
+        val reileigh = curG.joinVertices(v) {
+          case (_, x, y) => x * y
+        }.vertices.values.sum() / curG.vertices.mapValues(x => x * x).values.sum()
+
+        if (math.abs(norm - math.abs(reileigh)) > tol) {
+          throw new SparkException("Power Iteration fail to converge")
```

Comment:
Can you log anything else here in the exception message like the tolerance and norm, etc.?

Reply:
Yes. I would like to know, instead of throwing exception, can we just print  log error stating that it is failed to converge? But, in that case, we can't write UT.